cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsInstanceConnector.java

```diff
@@ -133,18 +133,52 @@ public String getConsoleOutput(AuthenticatedContext authenticatedContext, CloudI
     public List<CloudVmInstanceStatus> reboot(AuthenticatedContext ac, List<CloudInstance> vms) {
         AmazonEC2Client amazonEC2Client = awsClient.createAccess(new AwsCredentialView(ac.getCloudCredential()),
                 ac.getCloudContext().getLocation().getRegion().value());
+        List<CloudInstance> affectedVms = new ArrayList<>();
         try {
-            Collection<String> instances = vms.stream().map(CloudInstance::getInstanceId).collect(Collectors.toSet());
-            if (!instances.isEmpty()) {
-                amazonEC2Client.rebootInstances(new RebootInstancesRequest().withInstanceIds(instances));
+            if (!vms.isEmpty()) {
+                List<CloudVmInstanceStatus> statuses = check(ac, vms);
+                doReboot(affectedVms, amazonEC2Client, statuses.stream().filter(status -> status.getStatus() == InstanceStatus.STARTED)
+                        .map(status -> status.getCloudInstance()).collect(Collectors.toList()));
+                doStart(affectedVms, ac, statuses.stream().filter(status -> status.getStatus() == InstanceStatus.STOPPED)
+                        .map(status -> status.getCloudInstance()).collect(Collectors.toList()));
```

Comment:
could you move these stream parameters into variables or even to methods?

Reply:
moved